### DISCUSSION TITLE
Handling of CSRF tokens

### DISCUSSION THREAD

**Discussion by tmbenhura** [unpoly/unpoly#375] [Q&A]
### Invalid CSRF Token on Some Requests

I occasionally experience invalid CSF token errors from the backend application. I haven't been able to reproduce the circumstances that bring these errors about. I have a meta[name=csrf-token] tag in the header, as well as a hidden input in the form.

What is the recommended way of keeping meta tags up-to-date? There is unusual behaviour setting meta tags to up-hungry, and I suspect the meta tag is getting out of sync with the form csrf token.

### Expected behaviour

No invalid CSRF token errors

### Browser version

 - Unpoly 2.5

**Comment by adam12** [comment]
What backend are you pairing Unpoly with in this case?

**Comment by tmbenhura** [comment]
With Laravel

**Comment by triskweline** [comment]
At least in Rails this could happen if the user's session expired while the form was already loaded.

This case would be broken with or without Unpoly. You can keep the meta tag updated with `[up-hungry]` and/or `[up-poll]`.

**Reply by tmbenhura** [reply]
I think I have managed to replicate the problem, in this Glitch. The meta tag compiler seems to be evaluated last, though it is defined first. There is thus a mismatch if another compiler depends on the meta tag value.

https://glitch.com/edit/#!/nonchalant-cat-shade

**Reply by adam12** [reply]
Do you use a compiler for setting the CSRF token?

**Comment by tmbenhura** [comment]
The CRSF is updated by an up-hungry meta field on each request

It is then automatically picked when making ajax requests. These are
initiated within an on up:fragment:inserted in some circumstances.
